File: testdata/fixtures/ruby/CWE-89.rb
# CWE-89 SQL Injection in Ruby / ActiveRecord
# Building SQL via string interpolation with user input bypasses parameterisation
# and lets an attacker inject arbitrary SQL.

def search
  user = params[:user]
  name = params["name"]
  qs = request.query_string
  ck = cookies[:id]

  # ruleid: ruby-sql-injection
  ActiveRecord::Base.connection.execute("SELECT * FROM users WHERE name = '#{user}'")

  # ruleid: ruby-sql-injection
  User.where("name = '#{name}'")

  # ruleid: ruby-sql-injection
  User.find_by_sql("SELECT * FROM users WHERE id = #{qs}")

  # ruleid: ruby-sql-injection
  ActiveRecord::Base.connection.exec_query("SELECT * FROM users WHERE id = " + ck)

  # ok: ruby-sql-injection
  User.where("name = ?", "fixed")

  # ok: ruby-sql-injection
  ActiveRecord::Base.connection.execute("SELECT * FROM users WHERE name = 'fixed'")

  # ok: ruby-sql-injection
  User.find(1)
end

File: testdata/fixtures/rust/CWE-89.rs
// CWE-89: SQL Injection via sqlx query with untrusted input
//
// Passing user input directly as the SQL string lets an attacker inject
// arbitrary SQL. The sink is sqlx::query with tainted data.

fn vuln() {
    let input = std::env::var("USER_INPUT").unwrap();
    // ruleid: rust-sqli
    let _ = sqlx::query(&input);
    let data = std::env::args().next().unwrap();
    // ruleid: rust-sqli
    let _ = sqlx::query(&data);
    let propagated = input;
    // ruleid: rust-sqli
    let _ = sqlx::query(&propagated);
}

fn safe() {
    // ok: rust-sqli
    let _ = sqlx::query("SELECT * FROM users WHERE id=1");
    // ok: rust-sqli
    let _ = sqlx::query("SELECT * FROM t WHERE name='fixed'");
    // ok: rust-sqli
    let _ = sqlx::query("SELECT * FROM users");
}
